enhancements/installer/aws-lb-subnet-selection.md

```diff
@@ -0,0 +1,974 @@
+- Support configuring subnets for LoadBalancer-type Services that aren't associated with an IngressController.
+
+## Proposal
+
+To enable cluster admins to specify subnets for IngressControllers at install time, we will need
+to make the following API updates:
+
+1. Deprecate the `platform.aws.subnets` field in the install-config and replace with
+   `platform.aws.subnetsConfig`, which more intuitively manages multiple subnet roles.
+2. Add the `spec.loadBalancer.platform.aws.classicLoadBalancer.subnets` and
+   `spec.loadBalancer.platform.aws.networkLoadBalancer.subnets` fields in the IngressConfig
+   to encode the default subnets for IngressControllers.
+
+The proposal to deprecate the `platform.aws.subnets` field in the install-config is driven by
+its ambiguity, as the field name does not clearly distinguish between IngressController-role subnets and cluster-role
+subnets, which can be different. Additionally, `platform.aws.subnets` is a `[]string` and doesn't
+allow for future expansion or adaptation for other subnet type, roles, or metadata information.
+
+The IngressConfig holds the default values that the Ingress Operator will use when creating
+the `default` IngressController, as well as any new IngressController. The IngressConfig will be
+populated by the installer, similar to how the [Allow Users to specify Load Balancer Type during installation for AWS](/enhancements/installer/aws-load-balancer-type.md)
+enhancement introduced the `spec.loadBalancer.platform.aws.lbType` field.
+
+The OpenShift Installer will be updated to support the new `platform.aws.subnetsConfig` field, and the
+Ingress Operator will be updated to use the `Subnets` fields in the IngressConfig as the default subnets
+for new IngressControllers.
+
+### Implementation Details/Notes/Constraints
+
+As mentioned in [the Proposal section](#proposal), the install-config and the IngressConfig will be updated
+to support propagating default subnets to all new IngressControllers, including the `default`
+IngressController. The Ingress Operator will also need to be updated to use the new IngressConfig
+`Subnet` fields.
+
+#### Installer Updates
+
+The `platform.aws.subnets` field in the install-config will be deprecated and replaced with a new
+field, `platform.aws.subnetsConfig`. Much like `Subnets`, the new `SubnetsConfig` field indicates
+a list of subnets in a pre-existing VPC, but also provide the role that the subnet will fulfill in the
+cluster. Additionally, since it is a struct and not a `[]string`, it provide the ability to be
+expanded with additional subnet-related fields in the future.
+
+The following install-config reflect the initial TechPreview release of this feature:
+
+```go
+// Platform stores all the global configuration that all machinesets
+// use.
+// +kubebuilder:validation:XValidation:rule=`!(has(self.subnets) && has(self.subnetsConfig))`,message="cannot use both subnets and subnetsConfig"
```

Comment:
> Install config isn't a CRD,

FWIW we do generate an install config CRD for internal usage:
https://github.com/openshift/installer/blob/master/data/data/install.openshift.io_installconfigs.yaml

> do you think it would be useful to have the install config have the ability to leverage markers like other Kube APIs, then internally generate a json schema and run it through CRD validation? 

Yes it strikes me that it would be useful in terms of code readability & maintenance (not sure if you had other benefits in mind), as the installer validation code is difficult to maintain, it seems like CEL might be a better way to organize. 

That said, installer validation code is already so deeply developed it is unrealistic (and not worth the squeeze) to untangle it. I'm not sure how we would approach this problem: hybrid/frankenstein validation?

Reply:
In theory we could implement the CRD validation style as a new approach and have new additions validated using the new approach, and existing validations remain as they are (perhaps over time slowly migrating?)

> code readability & maintenance (not sure if you had other benefits in mind), as the installer validation code is difficult to maintain,

These were the main things I was thinking, and also consistency, people are used to writing API validations, so if the install config could follow a similar pattern, then the barrier to writing features touching the install config is reduced